### PR TITLE
infoschema/slow_query:  add cache for slow_query

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Before this PR, every operation on  table `slow_query` has to parse slow log file. 

This PR add a `ringBuffer` to cache the data of parse slow log.

### What is changed and how it works?

* This PR implement a simple `ringbuffer` to cache slow query data. The cache strategy is cache the latest slow log data. If the cache is full, use the latest data to cover the oldest data.

Below is a simple bench with 300M slow log file and the capacity of the cache is more than 300M.
before this PR:
```
always spend 2.3s
```

This PR:
```
First time: 2.3s
after: 0.5s
```

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

### CODE REVIEW

**Comment by crazycs520** [comment]
@zz-jason  @winkyao  @tiancaiamao  PTAL.

**Comment by crazycs520** [comment]
@qw4990 @winkyao PTAL

File: infoschema/slow_log_cache.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+	for i := range afterCacheTuples {
+		rows = append(rows, afterCacheTuples[i].convertToDatumRow())
+	}
+	// Update cache.
+	go s.updateCache(s.cache.filePath, afterCacheTuples)
+	return rows, nil
+}
+
+// parseSlowLogFileAndUpdateCache gets data by parse the slow log file, and then update the cache.
+func (s *slowLogReader) parseSlowLogFileAndUpdateCache(tz *time.Location, filePath string, offset int64, filterFn func(t time.Time) bool, bypassFn func(t time.Time) bool) ([][]types.Datum, error) {
```
Comment (qw4990) [inline infoschema/slow_log_cache.go]: `filterFn` and `bypassFn` are always nil, so why don't remove them?

File: infoschema/slow_log.go [hunk 2]
```diff
@@ -64,12 +64,32 @@ var slowQueryCols = []columnInfo{
...
 // TODO: Support parse multiple log-files.
-func parseSlowLogFile(tz *time.Location, filePath string) ([][]types.Datum, error) {
+func ReadSlowLogData(filePath string, tz *time.Location) ([][]types.Datum, error) {
+	return globalSlowQueryReader.getSlowLogData(filePath, tz)
+}
+
+// ParseSlowLogRows reads slow log data by parse slow log file.
+// It won't use the cache data.
+// It is exporting for testing.
+func ParseSlowLogRows(filePath string, tz *time.Location) ([][]types.Datum, error) {
```
Comment (qw4990) [inline infoschema/slow_log.go]: Since we have `ReadSlowLogData` now, why not remove this function?

File: infoschema/slow_log_cache.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+	cacheEndTuple := s.cache.getEndTuple()
+	for i := range tuples {
+		if tuples[i].time.Before(cacheEndTuple.time) {
+			continue
+		}
+		if tuples[i].equal(cacheEndTuple) {
+			continue
+		}
+
+		s.cache.buf.write(tuples[i])
```
Comment (qw4990) [inline infoschema/slow_log_cache.go]: Should we update `cacheEndTuple` after this statement?

Comment (crazycs520) [reply infoschema/slow_log_cache.go]: Good catch.
But normally, there won't have duplicate `tuple` in `tuples`, so I think it is no need to do this.

File: infoschema/slow_log.go (hunk 2 above)
Comment (crazycs520) [reply infoschema/slow_log.go]: `ParseSlowLogRows` only use to test and `ParseSlowLogRows` won't update cache.
`ReadSlowLogData` will update cache, but some test only wanto test parse slow log.

File: infoschema/slow_log_cache.go (hunk 1 above)
Comment (crazycs520) [reply infoschema/slow_log_cache.go]: great, thanks.

File: infoschema/slow_log.go [hunk 4]
```diff
@@ -64,12 +64,32 @@ var slowQueryCols = []columnInfo{
...
+// It is exporting for testing.
 // TODO: Support parse multiple log-files.
-func parseSlowLogFile(tz *time.Location, filePath string) ([][]types.Datum, error) {
+func ReadSlowLogData(filePath string, tz *time.Location) ([][]types.Datum, error) {
+	return globalSlowQueryReader.getSlowLogData(filePath, tz)
+}
+
+// ParseSlowLogRows reads slow log data by parse slow log file.
+// It won't use the cache data.
+// It is exporting for testing.
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply infoschema/slow_log.go]: Great, thanks.

File: infoschema/slow_log.go [hunk 5]
```diff
@@ -64,12 +64,32 @@ var slowQueryCols = []columnInfo{
 }
 
 func dataForSlowLog(ctx sessionctx.Context) ([][]types.Datum, error) {
-	return parseSlowLogFile(ctx.GetSessionVars().Location(), ctx.GetSessionVars().SlowQueryFile)
+	return ReadSlowLogData(ctx.GetSessionVars().SlowQueryFile, ctx.GetSessionVars().Location())
 }
 
-// parseSlowLogFile uses to parse slow log file.
+// ReadSlowLogData uses to read slow log data.
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply infoschema/slow_log.go]: Great, Thanks.

**Comment by qw4990** [comment]
> Is it a real demand to make the slow query check faster?
> Does it worth that we waste 300M memory to cache something not been frequently used?

Have the same question too. 🤔 
And there is a confusing point that if the user only needs the latest 300MB data, we can just parse the latest 300MB data which is fast enough I think, and if the user needs the whole slow log, the latest 300MB data cached in memory has no use.

**Comment by crazycs520** [comment]
@tiancaiamao How about adding a config to configure the buffer size? This demand has come from the client when they query slow_query was so slow.

**Comment by crazycs520** [comment]
@zz-jason PTAL

**Pull request closed without merging**